go/vt/vtgate/engine/concatenate.go

```diff
@@ -42,12 +42,20 @@ func (c *Concatenate) RouteType() string {
 
 // GetKeyspaceName specifies the Keyspace that this primitive routes to
 func (c *Concatenate) GetKeyspaceName() string {
-	return formatTwoOptionsNicely(c.Sources[0].GetKeyspaceName(), c.Sources[1].GetKeyspaceName())
+	res := c.Sources[0].GetKeyspaceName()
+	for i := 1; i < len(c.Sources); i++ {
+		res = formatTwoOptionsNicely(res, c.Sources[i].GetKeyspaceName())
```

Comment:
formatTwoOptionsNicely is not great for more than two options. I think you need to build a slice with the keyspaces and uniquefy it.

Reply:
GetKeyspaceName only returns a string. so if there are nested concatenate statements then it is not possible to eliminate all duplicates. for eg - 1 concatenate returns user_main and the other returns user_product, there is no way to find out that the user table is repeated between the two